vecindex: fix bugs found during testing of vector indexes

andy-kimball opened the pull request:
#### cspann: do not overwrite root level when searching targets

In the case where the root partition is in a non-Ready state, we need
to search its target partitions. Previously, we were overwriting the
root partition's level, which should be the same as the target partition's
level. However, if there's a bug, it can be different, and overriding
the root level causes cascading problems downstream, including a panic
that can take down the server. This commit adds code to check the level
and assert that it's equal to the root partition level rather than
silently overwriting it.

#### vecindex: clone treeKey when creating fixups

Previously, the C-SPANN index assumed that it owned the memory for
the vector index prefix value (called the "tree key" in C-SPANN).
However, the backfiller assumed the opposite, and reuses the memory
between SearchForInsert calls. This commit resolves the conflict by
always cloning the tree key when it's enqueued as part of a fixup,
so that callers can reuse the memory.

Fixes: #145261

#### cspann: pass tree key to delete vector fixup

Previously, the delete vector fixup did not take a tree key parameter,
which meant it always searched the default tree for the vector to
delete. This commit fixes that by passing the tree key to the fixup.

#### vecindex: return minimum number of results from Search

When vector deletion is enabled in TestVecIndexConcurrency, it occasionally
fails under stress with zero results. This is because searching with a LIMIT
of one was returning only 2 results. If both of those results point to deleted
primary key rows, then zero results get returned. The fix is to return at
least 10 candidate vectors from searches. While all 10 vectors could be deleted,
this at least makes this scenario much less common. The longer-term fix is to
switch to an iterator execution model, where we continue to "pull" additional
vectors until we get the desired number of results.

Release justification: Vector indexing is a business priority. These changes are all in the vecindex package, for a feature that is protected by a default-off feature flag.

andy-kimball committed:
cspann: pass tree key to delete vector fixup

Previously, the delete vector fixup did not take a tree key parameter,
which meant it always searched the default tree for the vector to
delete. This commit fixes that by passing the tree key to the fixup.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: do not overwrite root level when searching targets

In the case where the root partition is in a non-Ready state, we need
to search its target partitions. Previously, we were overwriting the
root partition's level, which should be the same as the root partition's
level. However, if there's a bug, it can be different, and overriding
the root level causes cascading problems downstream, including a panic
that can take down the server. This commit adds code to check the level
and assert that it's equal to the root partition level rather than
silently overwriting it.

Epic: CRDB-42943
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145274)
<!-- Reviewable:end -->

andy-kimball committed:
vecindex: clone treeKey when creating fixups

Previously, the C-SPANN index assumed that it owned the memory for
the vector index prefix value (called the "tree key" in C-SPANN).
However, the backfiller assumed the opposite, and reuses the memory
between SearchForInsert calls. This commit resolves the conflict by
always cloning the tree key when it's enqueued as part of a fixup,
so that callers can reuse the memory.

Fixes: #145261
Epic: CRDB-42943
Release note: None

andy-kimball committed:
vecindex: return minimum number of results from Search

When vector deletion is enabled in TestVecIndexConcurrency, it occasionally
fails under stress with zero results. This is because searching with a LIMIT
of one was returning only 2 results. If both of those results point to deleted
primary key rows, then zero results get returned. The fix is to return at
least 10 candidate vectors from searches. While all 10 vectors could be deleted,
this at least makes this scenario much less common. The longer-term fix is to
switch to an iterator execution model, where we continue to "pull" additional
vectors until we get the desired number of results.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Do you think this will fix #145299?

> Reviewed 3 of 3 files at r1, 6 of 6 files at r2, 3 of 3 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145274#-:-OOxNcTECpZH4FwFbic0:bf5sio6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145274#-OOxMZPz7nBhXcaUvblR:-OOxMZPz7nBhXcaUvblS:b-57bo1m):*
nit: `should be the same as the root partition's` -> `should be the same as the target partition's`


<!-- Sent from Reviewable.io -->

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 6 of 6 files at r2, 3 of 3 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145274#-:-OOxd-Vx19qLy-IbtQL6:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
